src/libmeasurement_kit/net/connect.cpp

```diff
@@ -171,33 +171,10 @@ void connect_ssl(bufferevent *orig_bev, ssl_st *ssl, std::string hostname,
                     return;
                 }
 
-                logger->debug("SSL version: %s", SSL_get_version(ssl));
-
-                long verify_err = SSL_get_verify_result(ssl);
-                if (verify_err != X509_V_OK) {
-                    debug("ssl: got an invalid certificate");
-                    bufferevent_free(bev);
-                    cb(SslInvalidCertificateError(
-                           X509_verify_cert_error_string(verify_err)),
-                       nullptr);
-                    return;
-                }
-
-                X509 *server_cert = SSL_get_peer_certificate(ssl);
-                if (server_cert == nullptr) {
-                    logger->debug("ssl: got no certificate");
-                    bufferevent_free(bev);
-                    cb(SslNoCertificateError(), nullptr);
-                    return;
-                }
-
-                hostname_validate_err =
-                    tls_check_name(nullptr, server_cert, hostname.c_str());
-                X509_free(server_cert);
-                if (hostname_validate_err != 0) {
-                    logger->debug("ssl: got invalid hostname");
+                err = libssl::verify_peer(hostname, ssl, logger);
+                if (err) {
                     bufferevent_free(bev);
-                    cb(SslInvalidHostnameError(), nullptr);
+                    cb(err, nullptr);
                     return;
```

Comment:
Block of code above factored as `libssl::verify_peer()`. We retain previous behavior of freeing `bev` on error.